iml-agent/src/daemon_plugins/devices/flat_devices.rs

```diff
@@ -0,0 +1,496 @@
+            device_type: x.device_type(),
+            parents: BTreeSet::new(),
+            children: x
+                .children
+                .iter()
+                .filter_map(|x| x.try_into_id().ok())
+                .collect(),
+            paths,
+            mount_path: x.mount.as_ref().map(|x| x.target.0.as_path()),
+            fs_type: x.mount.as_ref().map(|x| &x.fs_type.0),
+            fs_label: None,
+            fs_uuid: None,
+        })
+    }
+}
+
+impl<'a> TryFrom<&'a Dataset> for FlatDevice<'a> {
+    type Error = DeviceError;
+
+    fn try_from(x: &'a Dataset) -> Result<Self, Self::Error> {
+        let mut paths = BTreeSet::new();
+        paths.insert(Path::new(&x.name));
+
+        Ok(FlatDevice {
+            id: x.try_into_id()?,
+            size: 0,
+            device_type: x.device_type(),
+            parents: BTreeSet::new(),
+            children: BTreeSet::new(),
+            paths,
+            mount_path: x.mount.as_ref().map(|x| x.target.0.as_path()),
+            fs_type: x.mount.as_ref().map(|x| &x.fs_type.0),
+            fs_label: None,
+            fs_uuid: None,
+        })
+    }
+}
+
+fn convert_paths<'a>(xs: &'a device_types::devices::Paths) -> BTreeSet<&'a Path> {
+    xs.iter().map(|x| x.0.as_path()).collect()
+}
+
+pub type FlatDevices<'a> = BTreeMap<DeviceId, FlatDevice<'a>>;
+
+pub fn process_tree<'a>(
+    tree: &'a devices::Device,
+    parent_id: Option<DeviceId>,
+    ds: &mut FlatDevices<'a>,
+) {
+    match tree {
+        devices::Device::Root(Root { children }) => {
+            for c in children {
+                process_tree(c, None, ds)
+            }
+        }
+        devices::Device::ScsiDevice(x) => match FlatDevice::try_from(x) {
+            Ok(device) => {
+                let id = device.id.clone();
+
+                ds.entry(device.id.clone())
+                    .and_modify(|d| d.paths.extend(convert_paths(&x.paths).iter()))
+                    .or_insert(device);
+
+                for c in x.children.iter() {
+                    process_tree(c, Some(id.clone()), ds)
+                }
+            }
+            Err(e) => tracing::warn!("discarding ScsiDevice {:?} due to error {}", x, e),
+        },
+        devices::Device::Partition(x) => match FlatDevice::try_from(x) {
+            Ok(device) => {
+                let id = device.id.clone();
+
+                let device = ds
+                    .entry(device.id.clone())
+                    .and_modify(|d| d.paths.extend(convert_paths(&x.paths).iter()))
+                    .or_insert(device);
+
+                if let Some(p) = parent_id {
+                    device.parents.insert(p);
+                }
+
+                for c in x.children.iter() {
+                    process_tree(c, Some(id.clone()), ds)
+                }
+            }
+            Err(e) => tracing::warn!("discarding Partition {:?} due to error {}", x, e),
+        },
+        devices::Device::MdRaid(_) => {}
```

Comment:
This is todo or we're not doing anything with `MdRaid`s?

Reply:
todo.